pennylane/decomposition/controlled_decomposition.py

```diff
@@ -0,0 +1,239 @@
+            for w, val in zip(control_wires, control_values):
+                if not val:
+                    qml.PauliX(w)
+            qml.ctrl(
+                self._base_decomposition._impl,  # pylint: disable=protected-access
+                control=control_wires,
+                work_wires=work_wires,
+            )(*base.parameters, wires=base.wires, **base.hyperparameters)
+            for w, val in zip(control_wires, control_values):
+                if not val:
+                    qml.PauliX(w)
+
+        return _impl
+
+    def _get_resource_fn(self) -> Callable:
+        """The resource function."""
+
+        def _resource_fn(
+            *_, base_params, num_control_wires, num_zero_control_values, num_work_wires, **__
+        ):
+            base_resource_decomp = self._base_decomposition.compute_resources(**base_params)
+            gate_counts = {
+                _controlled_resource_rep(base_op_rep, num_control_wires, num_work_wires): count
+                for base_op_rep, count in base_resource_decomp.gate_counts.items()
+            }
+            gate_counts[resource_rep(qml.PauliX)] = num_zero_control_values * 2
+            return gate_counts
+
+        return _resource_fn
+
+
+def _controlled_g_phase_resource(
+    *_, num_control_wires, num_zero_control_values, num_work_wires, **__
+):
+    if num_control_wires == 1:
+        return {
+            resource_rep(qml.PauliX): num_zero_control_values * 2,
+            resource_rep(qml.PhaseShift): 1,
+        }
+
+    if num_control_wires == 2:
+        return {
+            resource_rep(qml.PauliX): num_zero_control_values * 2,
+            resource_rep(qml.ControlledPhaseShift): 1,
+        }
+
+    return {
+        resource_rep(qml.PauliX): num_zero_control_values * 2,
+        controlled_resource_rep(
+            qml.PhaseShift,
+            base_params={},
+            num_control_wires=num_control_wires - 1,
+            num_zero_control_values=0,
+            num_work_wires=num_work_wires,
+        ): 1,
+    }
+
+
+@register_resources(_controlled_g_phase_resource)
+def controlled_global_phase_decomp(*_, control_wires, control_values, work_wires, base, **__):
+    """The decomposition rule for a controlled global phase."""
+
+    for w, val in zip(control_wires, control_values):
+        if not val:
+            qml.PauliX(w)
+    if len(control_wires) == 1:
+        qml.PhaseShift(-base.data[0], wires=control_wires[-1])
+    else:
+        qml.ctrl(
+            qml.PhaseShift(-base.data[0], wires=control_wires[-1]),
+            control=control_wires[:-1],
+            work_wires=work_wires,
+        )
+    for w, val in zip(control_wires, control_values):
+        if not val:
+            qml.PauliX(w)
+
+
+def _controlled_x_resource(*_, num_control_wires, num_zero_control_values, num_work_wires, **__):
+    if num_control_wires == 1:
+        return {qml.CNOT: 1, qml.PauliX: num_zero_control_values * 2}
```

Comment:
Can do this with a single X for zero control (pulling the X through the control of the CNOT causes the two to cancel,while producing a single X on the target)

Reply:
Can you give an example?

Reply:
I meant like this:
```pycon
>>> mat_orig = qml.ctrl(qml.X(1), control=[0], control_values=[0]).matrix()
>>> mat_decomp = qml.matrix(qml.CNOT([0,1]) @ qml.X(1), wire_order=[0,1])
>>> np.allclose(mat_orig, mat_decomp)
True
```

Reply:
Basically, if there are two cases |0> and |1> for the control and we are "accidentally" flippping the target on |1> instead of |0>, we can just flip the target once more in any case, undoing the accidental flip on |1> and doing the intended flip on |0> :)